# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/QuotaCommon.h
# Commit: 0287f5604a69
# Full Hash: 0287f5604a69c65705743257a5bbe24a5bdb9d9f
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-12-14 23:03:37
# Regressor Bug: 1678030
# File Overlap Count: 1
# Description:
#   Bug 1678030 - Move CollectWhileHasResult to QuotaCommon and use in other quota clients. r=dom-workers-and-storage-reviewers,ttung
#   
#   This patch also:
#   
#   - introduces CollectElementsWhileHasResult(Typed) to collects elements in and
# ==============================================================================

diff -r 6dd8d66f8116 -r 0287f5604a69 dom/quota/QuotaCommon.h
--- a/dom/quota/QuotaCommon.h	Thu Dec 10 11:45:42 2020 +0000
+++ b/dom/quota/QuotaCommon.h	Mon Dec 14 10:42:36 2020 +0000
@@ -14,6 +14,7 @@
 #include <type_traits>
 #include <utility>
 #include "ErrorList.h"
+#include "mozIStorageStatement.h"
 #include "mozilla/Assertions.h"
 #include "mozilla/Attributes.h"
 #include "mozilla/Likely.h"
@@ -1117,6 +1118,39 @@
   return ExecuteSingleStep<ResultHandling>(std::move(stmt));
 }
 
+template <typename StepFunc>
+Result<Ok, nsresult> CollectWhileHasResult(mozIStorageStatement& aStmt,
+                                           StepFunc&& aStepFunc) {
+  return CollectWhile(
+      [&aStmt] { QM_TRY_RETURN(MOZ_TO_RESULT_INVOKE(aStmt, ExecuteStep)); },
+      [&aStmt, &aStepFunc] { return aStepFunc(aStmt); });
+}
+
+template <typename StepFunc,
+          typename ArrayType = nsTArray<typename std::invoke_result_t<
+              StepFunc, mozIStorageStatement&>::ok_type>>
+auto CollectElementsWhileHasResult(mozIStorageStatement& aStmt,
+                                   StepFunc&& aStepFunc)
+    -> Result<ArrayType, nsresult> {
+  ArrayType res;
+
+  QM_TRY(CollectWhileHasResult(
+      aStmt, [&aStepFunc, &res](auto& stmt) -> Result<Ok, nsresult> {
+        QM_TRY_UNWRAP(auto element, aStepFunc(stmt));
+        res.AppendElement(std::move(element));
+        return Ok{};
+      }));
+
+  return std::move(res);
+}
+
+template <typename ArrayType, typename StepFunc>
+auto CollectElementsWhileHasResultTyped(mozIStorageStatement& aStmt,
+                                        StepFunc&& aStepFunc) {
+  return CollectElementsWhileHasResult<StepFunc, ArrayType>(
+      aStmt, std::forward<StepFunc>(aStepFunc));
+}
+
 }  // namespace quota
 }  // namespace dom
 }  // namespace mozilla
